crates/mpz-zk/src/prover.rs

```
@@ -0,0 +1,291 @@
+            ctx.io_mut().send(uv).await?;
+        }
+
+        Ok(())
+    }
+}
+
+impl<OT> Vm<Binary> for Prover<OT>
+where
+    OT: RCOTReceiver<bool, Block>,
+{
+    type Error = Error;
+
+    fn call_raw(&mut self, call: Call) -> Result<Slice> {
+        let output = self.store.alloc_output(call.circ().output_len());
+
+        let mut count = call.circ().and_count();
+
+        // If the callstack is empty, we allocate more for the consistency check.
+        if self.callstack.is_empty() {
+            count += 128
+        }
+
+        self.ot.alloc(count).map_err(Error::ot)?;
+        self.callstack.push((call, output));
+
+        Ok(output)
+    }
+}
+
+impl<OT> Memory<Binary> for Prover<OT>
+where
+    OT: RCOTReceiver<bool, Block>,
+{
+    type Error = Error;
+
+    fn alloc_raw(&mut self, size: usize) -> Result<Slice> {
+        self.store.alloc_raw(size).map_err(Error::from)
+    }
+
+    fn assign_raw(&mut self, slice: Slice, data: BitVec) -> Result<()> {
+        self.store.assign_raw(slice, data).map_err(Error::from)
+    }
+
+    fn commit_raw(&mut self, slice: Slice) -> Result<()> {
+        self.store.commit_raw(slice).map_err(Error::from)
+    }
+
+    fn get_raw(&self, slice: Slice) -> Result<Option<BitVec>> {
+        self.store.get_raw(slice).map_err(Error::from)
+    }
+
+    fn decode_raw(&mut self, slice: Slice) -> Result<DecodeFuture<BitVec>> {
+        self.store.decode_raw(slice).map_err(Error::from)
+    }
+}
+
+impl<OT> View<Binary> for Prover<OT>
+where
+    OT: RCOTReceiver<bool, Block>,
+{
+    type Error = Error;
+
+    fn mark_public_raw(&mut self, slice: Slice) -> Result<()> {
+        self.store.mark_public_raw(slice).map_err(Error::from)
+    }
+
+    fn mark_private_raw(&mut self, slice: Slice) -> Result<()> {
+        self.store.mark_private_raw(slice)?;
+
+        self.ot.alloc(slice.len()).map_err(Error::ot)?;
+
+        Ok(())
+    }
+
+    fn mark_blind_raw(&mut self, slice: Slice) -> Result<()> {
```

Since there are no blind values for a zk prover, this should return an error.

---

https://github.com/privacy-scaling-explorations/mpz/pull/261